### PR TITLE
Add support for multiple top-level windows on Windows

### PR DESCRIPTION
**Pull request by hbatagelo**
Design doc: https://docs.google.com/document/d/1eQG-IS7r4_S9_h50MY_hSGwUVtgSTRSLzu7MLzPNd2Y/edit?tab=t.0

This PR is a part of a group of PRs. Going from first to last, these must be merged in the following order:

- https://github.com/flutter/engine/pull/56090
- https://github.com/flutter/engine/pull/56092
- https://github.com/flutter/engine/pull/56093
- https://github.com/flutter/engine/pull/56094

See also the corresponding framework PRs for multi-win support, starting from: https://github.com/flutter/flutter/pull/157515

## What's New

This pull request enables support for multiple top-level windows in Flutter applications on Windows. Specifically, this PR: 
- Adds the ` --enable-multi-window` command-line option, allowing applications to opt in to multi-window support (disabled by default).
- Introduces the `flutter/windowing` method channel, allowing windows to be created and destroyed from the framework.
  The following methods are available:
  - `createWindow(Map)`: Creates a top-level window.
    `Map` must contain a `<String, List<int>>` key-value pair, where the key is `"size"`, and the value is a list containing the requested width and height for the window's client rectangle, in logical coordinates.
    On success, a `Map` is returned with the following key-value pairs:
    | Key (`String`) | Value |
    | :--- | :--- |
    | `"viewId"` | The view ID of the window (`int`) |
    | `"archetype"` | `WindowArchetype::regular` |
    | `"size"` | The current size of the window (width, height), in logical coordinates (`List<int>`) |
    | `"parentViewId"` | `null` |

    If the method fails, a `PlatformException` is thrown.
  - `destroyWindow(Map)`: Destroys a window.
    `Map` must contain a `<String, int>` key-value pair, where the key is `"viewId"`, and the value is the view ID of the window to be destroyed. The method returns `void` on success. Otherwise, a `PlatformException` is thrown.

  The following callback methods can be listened to:
  - `onWindowCreated(Map)`: Called when a window is created.
     `Map` contains the following key-value pairs:
    | Key (`String`) | Value |
    | :--- | :--- |
    | `"viewId"` | The view ID (`int`) |
    | `"parentViewId"` | The parent view ID (`int`), or `null` if the window has no parent |
  - `onWindowDestroyed(Map)`: Called when a window is destroyed.
     `Map` contains a single `<String, int>` key-value pair, where the key is `"viewId"`, and the value is the view ID.
  - `onWindowChanged(Map)`: Called when a window property is changed (e.g., on resizing).
    `Map` contains the following key-value pairs:
    | Key (`String`) | Value |
    | :--- | :--- |
    | `"viewId"` | The view ID (`int`) |
    | `"size"` | The current size of the window in logical coordinates (`List<int>`), or `null` if the window size hasn't changed |

## How It Works

With multi-window support enabled, `FlutterWindowsEngine` creates a uniquely owned `FlutterHostWindowController` that is responsible for managing the lifecycle of windows associated with the engine. The engine registers the platform channel using `WindowingHandler`.

When the Dart app requests a new window using the `createX` methods, `WindowingHandler` calls `FlutterHostWindowController::CreateHostWindow` with the decoded arguments. The controller creates a uniquely owned `FlutterHostWindow` object which manages a native Win32 window with a child Flutter view in its client area. Each view, represented by `FlutterWindowsViewController`, is owned by its corresponding `FlutterHostWindow`. When the Dart app requests the destruction of a window (given a view ID) or when the native window is closed from the system, the controller destroys the corresponding `FlutterHostWindow`, which, in turn, cleans up the associated view controller.

Window messages are first handled by a static `FlutterHostWindow::WndProc` function that initializes the native Win32 window upon receiving `WM_NCCREATE`. Other messages are forwarded to `FlutterHostWindowController::HandleMessage` to manage global behaviors (e.g., hiding satellites of inactive top-level windows). Finally, messages are passed back to `FlutterHostWindow::HandleMessage` for window-specific behavior.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [X] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/windows/public/flutter_windows.h [hunk 1]
```diff
@@ -165,6 +174,22 @@ FLUTTER_EXPORT bool FlutterDesktopEngineDestroy(FlutterDesktopEngineRef engine);
...
+// |FlutterDesktopViewControllerDestroy|. Returns a null pointer in the event of
+// an error.
+//
+// Unlike |FlutterDesktopViewControllerCreate|, this does *not* take ownership
+// of |engine| and |FlutterDesktopEngineDestroy| must be called to destroy
+// the engine.
+FLUTTER_EXPORT FlutterDesktopViewControllerRef
+FlutterDesktopEngineCreateViewController(
+    FlutterDesktopEngineRef engine,
+    const FlutterDesktopViewControllerProperties* properties);
```
Comment (loic-sharma) [inline shell/platform/windows/public/flutter_windows.h:191]: FYI, we were holding off on making the multi-window APIs public until the Windows embedder properly supports multi-window.

Known problems with Window's multi-window support: https://flutter.dev/go/multi-window-status#heading=h.i5rwjhitgcmu

1. Cursor support
2. Keyboard support
3. Text input support
4. App lifecycle across multiple windows
5. Accessibility support
6. Plugins updated

Could we keep all these APIs non-public until we're sure these are production ready?

File: shell/platform/windows/client_wrapper/include/flutter/flutter_window_controller.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+#include <mutex>
+
+#include "flutter_engine.h"
+#include "method_channel.h"
+#include "win32_wrapper.h"
+#include "windowing.h"
+
+namespace flutter {
+
+// A singleton controller for Flutter windows.
```
Comment (loic-sharma) [inline shell/platform/windows/client_wrapper/include/flutter/flutter_window_controller.h:13]: Could you expand why this needs to be a singleton? What would I do if my app has two Flutter engines?

File: shell/platform/windows/client_wrapper/flutter_window_controller.cc [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+#include "include/flutter/encodable_value.h"
+#include "include/flutter/flutter_win32_window.h"
+#include "include/flutter/standard_method_codec.h"
+
+#include <iomanip>
+#include <sstream>
+
+#include <dwmapi.h>
+
+namespace {
```
Comment (loic-sharma) [inline shell/platform/windows/client_wrapper/flutter_window_controller.cc:12]: This `shell/platform/windows/client_wrapper` layer is code that the Flutter tool copies into the end user's project. This code doesn't get compiled into `flutter_windows.dll`. This doesn't seem like the right place to implement the `flutter/windowing` platform channel.

Our other platform channels are registered in the `FlutterWindowsEngine` [here](https://github.com/flutter/engine/blob/33f906426d067812215d46f8cad9f286ab2fdbab/shell/platform/windows/flutter_windows_engine.cc#L211-L226). Should we move the platform channel handling logic down to there?

File: shell/platform/windows/public/flutter_windows.h (hunk 1 above)
Comment (wanjm) [reply shell/platform/windows/public/flutter_windows.h:191]: I hope it can be public before the above 6 items are ready, so that more people can prototype this function; 
it is enough for flutter to just warn that it is not production ready;

Comment (loic-sharma) [reply shell/platform/windows/public/flutter_windows.h:191]: > I hope it can be public before the above 6 items are ready, so that more people can prototype this function; it is enough for flutter to just warn that it is not production ready;

The API is "pubternal"; advanced users can opt-in to using it (see https://github.com/flutter/flutter/issues/143767#issuecomment-2050741185). I think that's a good compromise as it prevents folks from unintentionally using an API that we know is broken.

**Comment by loic-sharma** [comment]
@hbatagelo would you be able to join the next Google / Canonical sync meeting to discuss this PR?

File: shell/platform/windows/client_wrapper/flutter_window_controller.cc (hunk 3 above)
Comment (hbatagelo) [reply shell/platform/windows/client_wrapper/flutter_window_controller.cc:12]: Understood. I'll move the handler to `FlutterWindowsEngine`

**Comment by hbatagelo** [comment]
> @hbatagelo would you be able to join the next Google / Canonical sync meeting to discuss this PR?

Sure. I'll be there.

**Comment by wanjm** [comment]
when will be the next sync meeting.

File: shell/platform/windows/client_wrapper/flutter_window_controller.cc [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+    return;
+  }
+
+  if (auto const data_opt{CreateFlutterWindow(
+          title, {.width = size_list->at(0), .height = size_list->at(1)},
+          archetype)}) {
+    auto const& data{data_opt.value()};
+    result.Success(EncodableValue(EncodableMap{
+        {EncodableValue("viewId"), EncodableValue(data.view_id)},
+        {EncodableValue("archetype"),
```
Comment (robert-ancell) [inline shell/platform/windows/client_wrapper/flutter_window_controller.cc:343]: Enumerated values in other channels are normally passed as strings, i.e. "Archetype.regular"

File: shell/platform/windows/client_wrapper/flutter_window_controller.cc [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+                                                 MethodCall<> const& call,
+                                                 MethodResult<>& result) {
+  auto const* const arguments{call.arguments()};
+  auto const* const map{std::get_if<EncodableMap>(arguments)};
+  if (!map) {
+    result.Error(kErrorCodeInvalidValue, "Method call argument is not a map.");
+    return;
+  }
+
+  std::wstring const title{ArchetypeToWideString(archetype)};
```
Comment (robert-ancell) [inline shell/platform/windows/client_wrapper/flutter_window_controller.cc:322]: Setting the title seems to the archetype seems like debugging code that shouldn't be here anymore?

File: shell/platform/windows/client_wrapper/flutter_window_controller.cc [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+
+  if (auto const data_opt{CreateFlutterWindow(
+          title, {.width = size_list->at(0), .height = size_list->at(1)},
+          archetype)}) {
+    auto const& data{data_opt.value()};
+    result.Success(EncodableValue(EncodableMap{
+        {EncodableValue("viewId"), EncodableValue(data.view_id)},
+        {EncodableValue("archetype"),
+         EncodableValue(static_cast<int>(data.archetype))},
+        {EncodableValue("size"),
```
Comment (robert-ancell) [inline shell/platform/windows/client_wrapper/flutter_window_controller.cc:345]: Might be easier to send "width" and "height" - less marshalling required.

File: shell/platform/windows/client_wrapper/flutter_window_controller.cc (hunk 4 above)
Comment (hbatagelo) [reply shell/platform/windows/client_wrapper/flutter_window_controller.cc:343]: Good to know, thanks. I'll change it.

File: shell/platform/windows/client_wrapper/flutter_window_controller.cc (hunk 5 above)
Comment (hbatagelo) [reply shell/platform/windows/client_wrapper/flutter_window_controller.cc:322]: Yes, it's just a placeholder. Setting the title is straightforward. I'll do it.

File: shell/platform/windows/client_wrapper/flutter_window_controller.cc (hunk 6 above)
Comment (hbatagelo) [reply shell/platform/windows/client_wrapper/flutter_window_controller.cc:345]: The current approach aims to maintain symmetry between the Dart API and the serialized structure, although we haven't given it much thought yet. I agree that a flat structure would simplify the serialization, but the nested approach may be more readable and intuitive as we add more attributes.

Maybe we should decide whether to adopt a flat structure (easier to serialize, but possibly cluttering the payload with many keys) or a nested structure (mimicking the Dart syntax but a bit more complex to parse) and write this down in the design doc.

Nested structure (using `createPopup` as an example):
```json
{
  "parentViewId": 42,
  "preferredSize": { "width": 800, "height": 600 },
  "anchorRect": { "left": 0, "top": 0, "width": 100, "height": 200 },
  "positionerParentAnchor": "topLeft",
  "positionerChildAnchor": "center",
  "positionerOffset": { "dx": 10, "dy": 20 },
  "positionerConstraintAdjustment": ["slideX", "slideY"]
}
```

Flat structure:
```json
{
  "parentViewId": 42,
  "preferredWidth": 800,
  "preferredHeight": 600,
  "anchorRectLeft": 0,
  "anchorRectTop": 0,
  "anchorRectWidth": 100,
  "anchorRectHeight": 200,
  "positionerParentAnchor": "topLeft",
  "positionerChildAnchor": "topLeft",
  "positionerOffsetX": 10,
  "positionerOffsetY": 20,
  "positionerConstraintAdjustment": ["slideX", "slideY"]
}
```

**Comment by hbatagelo** [comment]
Updates from https://github.com/canonical/flutter-engine/pull/1:

- The changes from the [multi-view patch on Windows](https://github.com/goderbauer/mvp/blob/main/patches/windows/001-Add-multi-view-Flutter-Windows-C%2B%2B-APIs.patch) have been reverted. The multi-window C API is now non-public again.
- The window controller, previously a singleton, is now instantiated once per engine and uniquely owned by it.
- Applications can opt in for multi-window support by using the `--enable-multi-window` command-line option.
- The engine now directly registers the platform channel for windowing communication. This channel is only registered for applications that opt in for multi-window support.
- Several fixes and improvements (prevented leakage of registered window classes, improved the behavior of modal dialogs by ensuring disabled windows do not appear active when selected via the task switcher, refactored tests, aligned the behavior of dialogs according to the spec, updated code style, improved comments).

**Comment by jtmcdole** [comment]
# Monorepo Migration Completed

> **TL;DR: Please migrate your PR to flutter/flutter**


The `flutter/engine` repository has been migrated to the `flutter/flutter` repository. This PR will no longer be landed here and must be migrated. To migrate your PR to the flutter repository, please follow these steps:

1. **Create a patch for this PR:**
   - Generate a patch set that represents the changes in this PR. The exact method will vary depending on your PR's history. **If your PR includes merges, it is highly recommended that you rebase it onto `main` first.**

   ```shell
   git format-patch $START_COMMIT..$END_COMMIT --stdout > combined_patch.patch
   ```

1. **Update the patch for the new engine location:**
   - The engine source code now resides in the `engine/` subdirectory within the `flutter/flutter` repository. You'll need to update the file paths in your patch accordingly.

   ```shell
   # Insert the `engine/` prefix to all paths. Note that this usage works on macOS and
   # linux versions of sed.
   sed -i.bak \
   -e 's|^\(diff --git a/\)\(.*\) b/\(.*\)|\1engine/\2 b/engine/src/flutter/\3|' \
   -e 's|^\(--- a/\)\(.*\)|\1engine/src/flutter/\2|' \
   -e 's|^\(\+\+\+ b/\)\(.*\)|\1engine/src/flutter/\2|' \
   combined_patch.patch
   ```

1. **Checkout and set up your Flutter development environment:**
   - Follow the instructions in [Setting up the Framework development environment](https://github.com/flutter/flutter/blob/master/docs/contributing/Setting-up-the-Framework-development-environment.md) to check out the `flutter/flutter` repository.
   - Remember to rename the origin remote to `upstream` (e.g., `git remote rename origin upstream`).

1. **Set up the engine development environment within the monorepo:**
   - Follow the updated instructions in [Setting up the engine environment](https://github.com/flutter/flutter/blob/master/engine/src/flutter/docs/contributing/Setting-up-the-Engine-development-environment.md), paying close attention to the changes in [gclient](https://github.com/flutter/flutter/blob/master/engine/README.md) setup and location.

1. **Apply the patch to a new branch:**
   - Create a new branch in your `flutter/flutter` repository.
   - Apply the updated patch to this branch:

   ```shell
   git apply combined_patch.patch
   ```

1. **Open a new PR:**
   - Open a new PR in the `flutter/flutter` repository with your changes.
   - Reference this original PR in the new PR's description using `flutter/engine#<PR_NUMBER>`.

___

*This is a canned message*
